### PR TITLE
Fix motchan types and checks

### PR DESCRIPTION
**Pull request by xezon**
I was looking through motchan once again because I still have another PR open on it... I saw a few things. Please check the commit section of this Pull Request to see each individual change applied and the rationale behind them.

### CODE REVIEW

**Comment by xezon** [comment]
This PR can be squashed when merged.

File: src/w3d/renderer/motchan.cpp [hunk 1]
```diff
@@ -385,11 +385,11 @@ bool TimeCodedBitChannelClass::Load_W3D(ChunkLoadClass &cload)
...
     m_defaultVal = chan.DefaultVal;
     m_cachedIdx = 0;
-    int bytesleft = 4 * m_numTimeCodes - 4;
+    const int bytesleft = 4 * m_numTimeCodes - 4;
 
-    captainslog_assert((sizeof(chan) + bytesleft) == (unsigned)chunk_size);
+    captainslog_assert((sizeof(chan) + bytesleft) == chunk_size);
 
-    m_bits = new unsigned long[m_numTimeCodes];
+    m_bits = new unsigned int[m_numTimeCodes];
```
Comment (jonwil) [inline src/w3d/renderer/motchan.cpp:405]: Confirmed as being unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 2]
```diff
@@ -54,23 +54,23 @@ class MotionChannelClass : public W3DMPO
...
+                setvec[i] = 0.0f;
             }
         }
     }
 
 private:
-    unsigned long m_pivotIdx;
-    unsigned long m_type;
+    int m_pivotIdx;
+    int m_type;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h]: These 2 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 3]
```diff
@@ -28,10 +28,10 @@ class MotionChannelClass : public W3DMPO
     void Free();
     bool Load_W3D(ChunkLoadClass &cload);
 
-    int Get_Type() { return m_type; }
-    int Get_Pivot() { return m_pivotIdx; }
+    int Get_Type() const { return m_type; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:31]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 4]
```diff
@@ -28,10 +28,10 @@ class MotionChannelClass : public W3DMPO
...
+    int Get_Pivot() const { return m_pivotIdx; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:32]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 5]
```diff
@@ -28,10 +28,10 @@ class MotionChannelClass : public W3DMPO
...
 
-    void Get_Vector(int frame, float *vector)
+    void Get_Vector(int frame, float *vector) const
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:34]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 6]
```diff
@@ -44,7 +44,7 @@ class MotionChannelClass : public W3DMPO
         }
     }
 
-    void Get_Vector_As_Quat(int frame, Quaternion &quat)
+    void Get_Vector_As_Quat(int frame, Quaternion &quat) const
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:47]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 7]
```diff
@@ -54,23 +54,23 @@ class MotionChannelClass : public W3DMPO
         }
     }
 
-    void Set_Identity(float *setvec)
+    void Set_Identity(float *setvec) const
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:57]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 8]
```diff
@@ -90,10 +90,10 @@ class BitChannelClass : public W3DMPO
     void Free();
     bool Load_W3D(ChunkLoadClass &cload);
 
-    int Get_Type() { return m_type; }
-    int Get_Pivot() { return m_pivotIdx; }
+    int Get_Type() const { return m_type; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:93]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 9]
```diff
@@ -90,10 +90,10 @@ class BitChannelClass : public W3DMPO
...
+    int Get_Pivot() const { return m_pivotIdx; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:94]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 10]
```diff
@@ -90,10 +90,10 @@ class BitChannelClass : public W3DMPO
...
 
-    int Get_Bit(int frame)
+    int Get_Bit(int frame) const
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:96]: Confirmed as const in the original.

File: src/w3d/renderer/motchan.h [hunk 11]
```diff
@@ -104,8 +104,8 @@ class BitChannelClass : public W3DMPO
     }
 
 private:
-    unsigned long m_pivotIdx;
-    unsigned long m_type;
+    int m_pivotIdx;
+    int m_type;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h]: These 2 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 12]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
     virtual ~TimeCodedMotionChannelClass() override;
     void Free();
     bool Load_W3D(ChunkLoadClass &cload);
-    int Get_Type() { return m_type; }
-    int Get_Pivot() { return m_pivotIdx; }
+    int Get_Type() const { return m_type; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:124]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 13]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
...
+    int Get_Pivot() const { return m_pivotIdx; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:125]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 14]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
...
-    unsigned long Binary_Search_Index(unsigned int timecode);
+    void Set_Identity(float *setvec) const;
+    unsigned int Get_Index(unsigned int timecode);
+    unsigned int Binary_Search_Index(unsigned int timecode) const;
 
 private:
-    unsigned long m_pivotIdx;
-    unsigned long m_type;
+    int m_pivotIdx;
+    int m_type;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h]: These 2 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 15]
```diff
@@ -148,17 +148,17 @@ class TimeCodedBitChannelClass : W3DMPO
     virtual ~TimeCodedBitChannelClass() override;
     void Free();
     bool Load_W3D(ChunkLoadClass &cload);
-    int Get_Type() { return m_type; }
-    int Get_Pivot() { return m_pivotIdx; }
+    int Get_Type() const { return m_type; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:151]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 16]
```diff
@@ -148,17 +148,17 @@ class TimeCodedBitChannelClass : W3DMPO
...
+    int Get_Pivot() const { return m_pivotIdx; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:152]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 17]
```diff
@@ -148,17 +148,17 @@ class TimeCodedBitChannelClass : W3DMPO
...
     int Get_Bit(int frame);
 
 private:
-    unsigned long m_pivotIdx;
-    unsigned long m_type;
+    int m_pivotIdx;
+    int m_type;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h]: These 2 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 18]
```diff
@@ -148,17 +148,17 @@ class TimeCodedBitChannelClass : W3DMPO
...
     int m_defaultVal;
-    unsigned long m_numTimeCodes;
-    unsigned long m_cachedIdx;
-    unsigned long *m_bits;
+    unsigned int m_numTimeCodes;
+    unsigned int m_cachedIdx;
+    unsigned int *m_bits;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:161]: These 3 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 19]
```diff
@@ -169,17 +169,17 @@ class AdaptiveDeltaMotionChannelClass : W3DMPO
...
     Quaternion Get_Quat_Vector(float frame_idx);
     float Get_Frame(unsigned int frame_idx, unsigned int vector_idx);
     void Decompress(unsigned int src_idx, float *srcdata, unsigned int frame_idx, float *outdata);
     void Decompress(unsigned int frame_idx, float *outdata);
 
 private:
-    unsigned int m_pivotIdx;
-    unsigned int m_type;
+    int m_pivotIdx;
+    int m_type;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h]: These 2 are confirmed as unsigned long in the original.

File: src/w3d/renderer/motchan.h [hunk 20]
```diff
@@ -169,17 +169,17 @@ class AdaptiveDeltaMotionChannelClass : W3DMPO
     virtual ~AdaptiveDeltaMotionChannelClass() override;
     void Free();
     bool Load_W3D(ChunkLoadClass &cload);
-    int Get_Type() { return m_type; }
-    int Get_Pivot() { return m_pivotIdx; }
+    int Get_Type() const { return m_type; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:172]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 21]
```diff
@@ -169,17 +169,17 @@ class AdaptiveDeltaMotionChannelClass : W3DMPO
...
+    int Get_Pivot() const { return m_pivotIdx; }
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:173]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 22]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
...
     void Get_Vector(float frame, float *setvec);
     Quaternion Get_Quat_Vector(float frame_idx);
-    void Set_Identity(float *setvec);
-    unsigned long Get_Index(unsigned int timecode);
-    unsigned long Binary_Search_Index(unsigned int timecode);
+    void Set_Identity(float *setvec) const;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:128]: This function isn't const in the original.

File: src/w3d/renderer/motchan.h [hunk 23]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
...
+    unsigned int Get_Index(unsigned int timecode);
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:129]: Confirmed in the original as returning unsigned long.

File: src/w3d/renderer/motchan.h [hunk 24]
```diff
@@ -121,17 +121,17 @@ class TimeCodedMotionChannelClass : W3DMPO
...
+    unsigned int Binary_Search_Index(unsigned int timecode) const;
```
Comment (jonwil) [inline src/w3d/renderer/motchan.h:130]: Confirmed in the original as returning unsigned long.
Also this function isn't const in the original.

**Review by jonwil — COMMENTED** [review]
Matched some things against the original and added notes based on what I found.

**Review by tomsons26 — CHANGES_REQUESTED** [review]
Problem with touching any of this is that means you need to regression test all animations in the game
Not something i want to deal with right now, i will check this PR later at some point and see what can be changed safely so leave it open for now

**Comment by xezon** [comment]
Regarding the const functions: it is just a compiler flag here telling compiler to generate error on non-const class member access. It does not change functionality. That is safe.

Regarding the `unsigned long` vs `int` - I would like to point out that the original code uses a mess of types. See below.

### Example 1

```c++
    MotionChannelClass::unsigned long m_pivotIdx;
    MotionChannelClass::unsigned long m_type;
    
    MotionChannelClass::int Get_Type() { return m_type; }
    MotionChannelClass::int Get_Pivot() { return m_pivotIdx; }
```
Note how these types are stored as `unsigned long` in class but returned as `int`.

### Example 2

```c++
TimeCodedMotionChannelClass::unsigned int m_cachedIdx
TimeCodedBitChannelClass::unsigned long m_cachedIdx
```
One class uses `unsigned int` and the other `unsigned long` for the same kind of thing.

### Example 3

```c++
TimeCodedMotionChannelClass::unsigned long Get_Index(unsigned int timecode)
```
is used as
```c++
void TimeCodedMotionChannelClass::Get_Vector(float frame, float *setvec)
{
    int index = Get_Index(frame);
}
```
`unsigned long` is used as `int`, and `float` is used as `unsigned int`.

This odd type mess goes on and on. It's not good.

File: src/w3d/renderer/motchan.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/w3d/renderer/motchan.cpp:405]: ~~strongly suggest changing all these to `uint64_t` as then its much easier to read.~~

**Comment by xezon** [comment]
### Regarding unsigned long m_type

It should be int, because the enum values are integer too. (enum is int by default)

```c++
enum
{
    ANIM_CHANNEL_X = 0,
    ANIM_CHANNEL_Y,
    ANIM_CHANNEL_Z,
    ANIM_CHANNEL_XR,
    ANIM_CHANNEL_YR,
    ANIM_CHANNEL_ZR,
    ANIM_CHANNEL_Q,

    ANIM_CHANNEL_TIMECODED_X,
    ANIM_CHANNEL_TIMECODED_Y,
    ANIM_CHANNEL_TIMECODED_Z,
    ANIM_CHANNEL_TIMECODED_Q,

    ANIM_CHANNEL_ADAPTIVEDELTA_X,
    ANIM_CHANNEL_ADAPTIVEDELTA_Y,
    ANIM_CHANNEL_ADAPTIVEDELTA_Z,
    ANIM_CHANNEL_ADAPTIVEDELTA_Q,
};

enum
{
    ANIM_FLAVOR_TIMECODED = 0,
    ANIM_FLAVOR_ADAPTIVE_DELTA,
    ANIM_FLAVOR_VALID
};
```

### Regarding unsigned long m_pivotIdx

It is used in int context here:

```c++
                if (channel->Get_Pivot() >= m_numNodes) {
                    delete channel;
                }
```

File: src/w3d/renderer/motchan.cpp (hunk 1 above)
Comment (xezon) [reply src/w3d/renderer/motchan.cpp:405]: `long` is 32 bit though.

Comment (duncanspumpkin) [reply src/w3d/renderer/motchan.cpp:405]: https://github.com/TheAssemblyArmada/Thyme/wiki/Thyme-Style-Guide also as per the style guide we should be avoiding it

Comment (xezon) [reply src/w3d/renderer/motchan.cpp:405]: I am ok with making it `uint32_t`

**Comment by duncanspumpkin** [comment]
I'm in agreement with xezon. We should be using the sized types for this anyway which would be `uint32_t`/`int32_t` then it is completely unambiguous.

**Review by tomsons26 — CHANGES_REQUESTED** [review]
revert the signage changes
HRawAnimClass::Get_Transform will actually need a deeper look, need to figure out where the best info source is for this one, but at a glance what you fix is actually as it should be, there's possibly code in wrong branches or out right missing in this function

**Comment by xezon** [comment]
Ok. pivot index and type changed to unsigned again.

**Comment by xezon** [comment]
Submit before this one:
* #547

**Comment by xezon** [comment]
I tested various units and buildings and animations looked as usual. No problems found.

**Comment by xezon** [comment]
Anything else that needs fixing here?

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**